Inc/HALAL/Services/Communication/I2C/I2C.hpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * UART.hpp
+ *
+ *  Created on: 12 dec. 2022
+ *      Author: Pablo
+ */
+#pragma once
+
+#include "PinModel/Pin.hpp"
+#include "C++Utilities/CppUtils.hpp"
+#include "Packets/I2CPacket.hpp"
+#include "ErrorHandler/ErrorHandler.hpp"
+#include "DMA/DMA.hpp"
+
+#define HALAL_I2C_EMPTY_ID (uint16_t)0
+
+#ifdef HAL_I2C_MODULE_ENABLED
+
+/**
+ * @brief I2C service class. Abstracts the use of the UART service of the HAL library.
+ * 
+ */
+class I2C{
+private:
+    /**
+     * @brief Struct which defines all data referring to I2C peripherals. It is
+     *        declared private in order to prevent unwanted use. Only 
+     *        predefined instances should be used.
+     *           
+     */
+    struct Instance{
+        Pin SCL; /**< Clock pin. */
+        Pin SDA; /**< MOSI pin. */
+        I2C_HandleTypeDef* hi2c;  /**< HAL UART struct. */
+        I2C_TypeDef* instance;
+        DMA::Stream RX_DMA;
+        DMA::Stream TX_DMA;
+        uint8_t address = 0;
+        uint32_t speed_frequency = 100; /**< frequency in khz*/
+        uint32_t data_length = 8; /**< size in bits*/
+        bool receive_ready = false; /**< Receive value is ready to use pin. */
+        bool initialized = false;
+
+    };
+
+public:
+    /**
+      * @brief Enum which abstracts the use of the Instance struct to facilitate the mocking of the HALAL.Struct
+      *
+      */
+     enum Peripheral{
+         peripheral2 = 2,
+     };
+
+public:
+    static uint16_t id_counter;
+
+    static unordered_map<uint8_t, I2C::Instance* > active_i2c;
+
+    static unordered_map<I2C::Peripheral, I2C::Instance*> available_i2cs;
+
+    static unordered_map<uint32_t, uint32_t> available_speed_frequencies;
+
+    /**
+     * @brief I2C 2 wrapper enum of the STM32H723.
+     *
+     */
+    static I2C::Peripheral i2c2;
+
+    static DMA_HandleTypeDef hdma_i2c2_rx;
+    static DMA_HandleTypeDef hdma_i2c2_tx;
+
+    /**
+     * @brief I2C 2 instance of the STM32H723.
+     *
+     */
+    static I2C::Instance instance2;
+
+    /**
+     * @brief Registers a new I2C.
+     * 
+     * @param uart I2C peripheral to register.
+     * @return uint8_t Id of the service.
+     */
+    static optional<uint8_t> inscribe(I2C::Peripheral& i2c);
+    static optional<uint8_t> inscribe(I2C::Peripheral& i2c, uint8_t address);
+
+    /**
+     * @brief This method initializes all registered I2Cs. The peripherals
+     * 		  must be enrolled before calling this method.
+     * 
+     */
+    static void start();
+
+    /**@brief	Transmits 1 RawPacket of any size by DMAs.
+     * 			Handles the packet size automatically.
+     * @param id Id of the I2C
+     * @param packet Packet to be send
+     * @return bool Returns true if the request to send the packet has been done
+     *            successfully. Returns false if the I2C is busy or a problem
+     *            has occurred.
+     */
+    static bool transmit_next_packet(uint8_t id, I2CPacket& packet);
```

Comment:
```suggestion
    static bool transmit(uint8_t id, I2CPacket& packet);
```
Que aporta next_packet?